opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java

```diff
@@ -518,24 +528,32 @@ boolean isNestable() {
         return true;
     }
 
+    private static boolean isGitWorking(String repoCommand) {
+        Executor exec = new Executor(new String[]{repoCommand, "--version"});
+        if (exec.exec(false) == 0) {
+            final String outputVersion = exec.getOutputString();
+            final String version = outputVersion.replaceAll(".*? version (\\d+(\\.\\d+)*).*", "$1");
+            try {
+                return Version.from(version).compareTo(MINIMUM_VERSION) >= 0;
+            } catch (NumberFormatException ex) {
+                LOGGER.log(Level.WARNING, String.format("Unable to detect git version from %s", outputVersion), ex);
+            }
+        }
+        return false;
+    }
+
     @Override
-    public boolean isWorking() {
+    public synchronized boolean isWorking() {
```

Comment:
How about double-locking inside the method (with proper `volatile working`) as the lowest-cost synchronization vs always synchronizing?

(I would think this affects all `RepositoryInfo` sub-classes but as just a mostly-harmless, occasional redundant-run. Or am I misunderstanding?)

Reply:
It seems that double-locking is an overkill here. Once GIT_IS_WORKING is initialized by one instance of GitRepository, the method will be very quick and lock contention is not likely to be an issue. Also, there are no other synchronized methods just yet. In general, inheritance from RepositoryInfo does not look like a good choice here, as it is mutable and not properly synchronized. Repository should be an interface, with, maybe a base abstract implementation class, allowing for more flexibility in how to make it thread-safe (immutability, etc)

I have not looked closely at other Repository subclasses, iirc, some of them actually run commands pertaining to the actual repository in order to determine whether a given Repository instance is working. I am trying to contribute back to the community some of the optimizations I ended up doing for my employer where we use opengrok with git, hence, only fixing GitRepository. This fix is useful to improve indexing performance of installations with many (e.g. 10k) repositories, so it is harmless from correctness perspective but not from indexing speed perspective.

Reply:
Ah OK I'm clear now that you're striving for once-per-run initialization of Git is-working. The pattern in OpenGrok would be to add a `static LazilyInstantiate<Boolean>` that synchronizes until initialized and thereafter doesn't bother to synchronize at all.

Reply:
@azakkerman, apologies if already answered, but you write the following:

> ... trying to contribute back to the community some of the optimizations I ended up doing for my employer ....

and on [another issue](https://github.com/oracle/opengrok/issues/3306):

> it requires some work to get past our corporate process for submitting OSS patches.

Would these require your company copyright to be asserted in contributions?

Reply:
> Ah OK I'm clear now that you're striving for once-per-run initialization of Git is-working. The pattern in OpenGrok would be to add a static LazilyInstantiate<Boolean> that synchronizes until initialized and thereafter doesn't bother to synchronize at all.

Updated the code

> Would these require your company copyright to be asserted in contributions?

I received approval. They ask to include work email in addition to personal email. Will double-check if they want to add copyright as well. The issue is mostly about exporting patches from the internal network, but I chose to simply re-implement the code.

Reply:
LazilyInstantiate only synchronizes the static GIT_IS_WORKING, but not this.working. Java Memory Model does not guarantee any visibility for read after write across threads w/o use of synchronizing primitives to force memory barriers. My assertion is that synchronized here is needed for correctness, unless you want to make working field of RepositoryInfo volatile. If you insist, I will remove synchronized, as, in the worst case, this would result in an extra GIT_IS_WORKING.get(), so it should be idempotent.

Reply:
Indeed. I mean since the reliance now is on `LazilyInstantiate` for synchronization, and `GIT_IS_WORKING.get()` can only ever return the same value, there is no need to synchronize here (nor to make `working` volatile) because redundant `get()` or assignment to `working` are harmless.